L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -112,27 +142,73 @@ superTriggerCellSelectImpl(const std::vector<l1t::HGCalTriggerCell>& trigCellVec
+  std::vector<l1t::HGCalTriggerCell> trigCellVecInputEnlarged;
 
-  // first pass, fill the super trigger cells
+  // first pass, fill the "coarse" trigger cells
   for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
-    if (tc.subdetId() == HGCHEB) continue;
-    STCs[getSuperTriggerCellId(tc.detId())].add(tc);
+    if (tc.subdetId() == HGCHEB){
+      trigCellVecInputEnlarged.push_back( tc );
+      continue;
+    }
+    int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+    int stcid = coarseTCmapping_.getCoarseTriggerCellId(tc.detId(),stcSize_.at(thickness));
+    STCs[stcid].add(tc, stcid);
   }
-    
+
+   //The missing trigger cells are needed to be created both for coarsening the existing TC
+  // in the thick modules (for the fixed data size choice), or for any of the energy spread algorithms (except super TCs)
+
+  createMissingTriggerCells( STCs, trigCellVecInputEnlarged);
+
+  //Change coarse TC positions if needed
+  if ( fixedDataSizePerHGCROC_ == true ){
+    for  (l1t::HGCalTriggerCell & tc : trigCellVecInputEnlarged) {
+      int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+      if ( thickness > 0 ){
+	coarseTCmapping_.setCoarseTriggerCellPosition( tc );
+      }
+    }
+  }
+
+  if ( energyDivisionType_ == oneBitFraction ){
+    //Get the 1 bit fractions. There should be exactly 4 TCs per STC
+    for (const l1t::HGCalTriggerCell & tc : trigCellVecInputEnlarged) {
+      if (tc.subdetId() == HGCHEB) continue;
+      int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+      
+      int stcid = coarseTCmapping_.getCoarseTriggerCellId(tc.detId(),stcSize_.at(thickness)); 
+      float stc_sumpt = STCs[stcid].GetSumPt();
+      float tc_fraction = getTriggerCellOneBitFraction( tc.pt() , stc_sumpt );
+      if ( tc.detId() != STCs[stcid].GetMaxId() ){
+	STCs[stcid].addToFractionSum(tc_fraction);
+      }
+    }
+  }
+
   // second pass, write them out
-  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
-    
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInputEnlarged) {
     //If scintillator use a simple threshold cut
     if (tc.subdetId() == HGCHEB) {
       trigCellVecOutput.push_back( tc );
     } else {
-      const auto & stc = STCs[getSuperTriggerCellId(tc.detId())]; 
-      if (tc.detId() == stc.GetMaxId() ) {
+      int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+      const auto & stc = STCs[coarseTCmapping_.getCoarseTriggerCellId(tc.detId(),stcSize_.at(thickness))]; 
+    
+      if ( (energyDivisionType_!=superTriggerCell && energyDivisionType_!=coarse2TriggerCell)
+           || ( energyDivisionType_==superTriggerCell && (tc.detId() == stc.GetMaxId()) )
+           || ( energyDivisionType_==coarse2TriggerCell && (!(tc.detId() & 1))  )
+           )  {
+
         trigCellVecOutput.push_back( tc );
-        stc.assignEnergy(trigCellVecOutput.back());
+        assignSuperTriggerCellEnergy(trigCellVecOutput.back(), stc);        
```

Comment:
Same type of comment as before. Can it be done directly during the creation of the output trigger cells?

Reply:
For all other than the 1 bit fraction, which needs another loop over the TCs to calculate the fraction sum. So maybe it's better to keep it how it is?

Reply:
Yes, so it can be a second TC loop. And `trigCellVecInputEnlarged` shouldn't be needed, `trigCellVecOutput` can be used directly to avoid copying trigger cells.

Reply:
Ok done, with the inclusion of a map in the STC class storing the original TC pts